File: Training/stack/stack.c
// // Online C compiler to run C program online
// #include <iostream>

// using namespace std;

// class Stack {
// public:
//     int *arr;
//     int top;
//     int size;

//     Stack(int size) {
//         this->size = size;
//         arr = new int[size];
//         top = -1;
//     }

//     void push(int element) {
//         if (top < size - 1) {
//             top++;
//             arr[top] = element;
//         } else {
//             cout << "Stack overflow" << endl;
//         }
//     }

//     void pop() {
//         if (top >= 0) {
//             top--;
//         } else {
//             cout << "Stack underflow" << endl;
//         }
//     }

//     int peek() {
//         if (top >= 0 && top < size)
//             return arr[top];
//         else {
//             cout << "Stack is empty" << endl;
//             return -1;
//         }
//     }

//     bool isEmpty() {
//         return top == -1;
//     }
// };

// int main() {
//     // Write C++ code here
//     Stack st(5);
//     st.push(22);
//     st.push(43);

//     cout << st.peek() << endl;
//     st.pop();

//     if (st.isEmpty()) {
//         cout << "Stack is empty" << endl;
//     } else {
//         cout << "Stack is not empty" << endl;
//     }

//     return 0;
// }

File: Training/stack/stack.go
package main

import (
	"errors"
	"fmt"
)

var arr [3]int
var top = -1
var size = len(arr)

func main() {
	fmt.Println("Is the stack empty?", isEmpty())
	fmt.Println("the first element is :", peak())
	push(10)
	push(20)
	push(30)
	fmt.Println("the first element is :", peak())
	fmt.Println("The element inside array is:")
	ArrElement(arr[:])
	poppedElement, err := pop()
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println("Popped element:", poppedElement)
	}
	push(2)
	fmt.Println("The element inside array is:")
	ArrElement(arr[:])

}

func push(element int) {
	if top < size-1 {
		top++
		arr[top] = element
	} else {
		fmt.Println("The stack is full , Stack overflow")
	}
}

func pop() (int, error) {
	if top == -1 {
		return -1, errors.New("THE stack is empty")
	} else {
		var res = arr[top]
		top--
		return res, nil
	}
}

func peak() int {
	if top >= 0 && top < size {
		return arr[top]
	} else {
		fmt.Println("The stack is empty")
		return -1
	}
}

func isEmpty() bool {
	return top == -1
}

func ArrElement(arr []int) {
	for _, val := range arr {
		fmt.Println("THE VALUE INSIDE IS", val)
	}
}
